src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/CompletionItemSource.cs

```diff
@@ -0,0 +1,249 @@
+
+            var service = document.GetLanguageService<CompletionService>();
+            if (service == null)
+            {
+                return EditorCompletion.CompletionStartData.DoesNotParticipateInCompletion;
+            }
+
+            var sourceText = document.GetTextSynchronously(cancellationToken);
+
+            // TODO: Check CompletionOptions.TriggerOnTyping  https://github.com/dotnet/roslyn/issues/27427
+            if (!(trigger.Reason == EditorCompletion.CompletionTriggerReason.Invoke ||
+                trigger.Reason == EditorCompletion.CompletionTriggerReason.InvokeAndCommitIfUnique)
+                && !service.ShouldTriggerCompletion(sourceText, triggerLocation.Position, GetRoslynTrigger(trigger, triggerLocation)))
+            {
+                return EditorCompletion.CompletionStartData.DoesNotParticipateInCompletion;
+            }
+
+            return new EditorCompletion.CompletionStartData(
+                participation: EditorCompletion.CompletionParticipation.ProvidesItems,
+                applicableToSpan: new SnapshotSpan(triggerLocation.Snapshot, service.GetDefaultCompletionListSpan(sourceText, triggerLocation.Position).ToSpan()));
+        }
+
+        public async Task<EditorCompletion.CompletionContext> GetCompletionContextAsync(
+            IAsyncCompletionSession session, 
+            EditorCompletion.CompletionTrigger trigger, 
+            SnapshotPoint triggerLocation, 
+            SnapshotSpan applicableToSpan, 
+            CancellationToken cancellationToken)
+        {
+            var document = triggerLocation.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return new EditorCompletion.CompletionContext(ImmutableArray<EditorCompletion.CompletionItem>.Empty);
+            }
+
+            var completionService = document.GetLanguageService<CompletionService>();
+
+            var completionList = await completionService.GetCompletionsAsync(
+                document,
+                triggerLocation,
+                GetRoslynTrigger(trigger, triggerLocation)).ConfigureAwait(false);
+
+            if (completionList == null)
+            {
+                return new EditorCompletion.CompletionContext(ImmutableArray<EditorCompletion.CompletionItem>.Empty);
+            }
+
+            var filterCache = new Dictionary<string, EditorCompletion.CompletionFilter>();
+
+            var itemsBuilder = ImmutableArray.CreateBuilder<EditorCompletion.CompletionItem>();
+            foreach (var roslynItem in completionList.Items)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                var item = Convert(document, roslynItem, completionService, filterCache);
+                item.Properties.AddProperty(TriggerBuffer, triggerLocation.Snapshot.TextBuffer);
+                if (!string.IsNullOrEmpty(item.DisplayText))
+                {
+                    itemsBuilder.Add(item);
```

Comment:
needs comment explaining why this is correct.

Reply:
1. remove the if-statement just above
2. other than the if-statement (and adding a property), this is just a conversion from Roslyn format to VS format.

---
In reply to: [222082424](https://github.com/dotnet/roslyn/pull/29016#discussion_r222082424) [](ancestors = 222082424)